design-proposals/data-import-cron-http-import.md

```diff
@@ -0,0 +1,78 @@
+
+# Overview
+This design provides an approach to allow DataImportCron to import HTTP sources.  
+The design will also demonstrate to the user how to operate with such import types.  
+
+## Motivation
+Currently, the DataImportCron allows the import of Registry Imports only. Recently, there was a demand to also allow HTTP import types.  
+The problem that arises from HTTP imports is that there is no convention between the different sources, so it's hard to know when the image is updated for each source in a generic way, which will make the polling process more difficult than standard registry sources.  
+One possible solution If the URL is static is to use a Get request with an If-Modified-Since header where we specify the date from which we want to check if there was a change. If there was a change since the specified date, the request will return with a status of 200OK, and then we know that the image has been updated since that date.  
+
+In the current situation we can enable HTTP imports by updating the DesiredDigest label.  
+
+## Goals
+
+* Allow the user to perform http imports manually with dataimportcron
+* Create a poller that will cover as many cases as possible for automatic updating
+
+## Non Goals
+
+* The poller will probably not cover all import cases and sometimes the user will have to manually update the digest
+
+## User Stories
+
+* As a user, I would like to import images from an HTTP source using DataImportCron
+* As a user, I would like the poller automatically update the image when it is needed
+* As a user, I would like to manually trigger an HTTP import with DataImportCron
+
+## Repos
+
+* **CDI**: Offers an API which serves as a mutable symbolic link to a Persistent Volume Claim (PVC). This link can be used in a DataVolume.
+* **CDI**: Provides an API called DataImportCron, designed to manage the periodic polling and importing of disk images as PVCs into a specific namespace (golden images by default). This functionality leverages the CDI import processes.
+
+## Implementation Phases
+
+**Phase1** - Manual Import Update
+* DataImportCron controller to handle HTTP imports
+* DataImportCron webhook/validation to support HTTP source type
+
+**Phase2** - Poller
+* DataImportCron poller to support HTTP source polling
+
+# Design
+
+If the URL is static:
+* Make a GET request with the If-Modified-Since Header starting from the date stored in the AnnLastCronTime annotation
+* If the returned status is 200OK, perform import again
+* Update AnnLastCronTime to time.Now()
```

Comment:
Have you considered that some sources provide a separate file with the hash of the image? For instance the fedora [cloud images](https://download.fedoraproject.org/pub/fedora/linux/releases/38/Cloud/x86_64/images/Fedora-Cloud-Base-38-1.6.x86_64.raw.xz) have a [file](https://download.fedoraproject.org/pub/fedora/linux/releases/38/Cloud/x86_64/images/Fedora-Cloud-38-1.6-x86_64-CHECKSUM) that contains the checksum of the file. So we could download that file and verify the hash against what we have and decide if a new one exists that way.

Reply:
Good point. The question is what percentage of sources support this and whether all sources provide the file in the same format?
For example https://cloud.debian.org/images/cloud/bookworm/daily/latest/SHA512SUMS looks different from the file you mentioned

Reply:
As far as I can tell these are the areas where this file structure diverges:
- Hash algorithm (MD5 in tinycore, SHA512 in debian, SHA256 in Fedora)
- File structure (`filename  hash` in most but fedora does `SHA256 (filename) = hash`)

Reply:
Maybe we can catalog these, if there is a limited number of permutations of these formats, maybe we can implement them, and only support those limited number of options. I understand there might be many, but if we capture the most common ones, that might be enough.